ImperatorToCK3/Source/CK3/CK3World.cpp

```diff
@@ -0,0 +1,170 @@
+					continue;
+				}
+				if (provinces.count(id))
+				{
+					Log(LogLevel::Info) << "Vanilla province duplication - " << id << " already loaded! Preferring unique entry over mapping.";
+				}
+				else
+					provinces.insert(std::pair(id, provinces.find(newProvince.second)->second));
+			}
+		}
+		catch (std::exception& e)
+		{
+			Log(LogLevel::Warning) << "Invalid province filename: " << ck3Path << "/game/history/province_mapping/" << fileName << " : " << e.what();
+		}
+	}
+	LOG(LogLevel::Info) << ">> Loaded " << provinces.size() << " province definitions.";
+}
+
+void CK3::World::importImperatorProvinces(const ImperatorWorld::World& sourceWorld)
+{
+	LOG(LogLevel::Info) << "-> Importing Imperator Provinces";
+	auto counter = 0;
+	// Imperator provinces map to a subset of CK3 provinces. We'll only rewrite those we are responsible for.
+	for (const auto& province : provinces)
+	{
+		const auto& impProvinces = provinceMapper.getImperatorProvinceNumbers(province.first);
+		// Provinces we're not affecting will not be in this list.
+		if (impProvinces.empty())
+			continue;
+		// Next, we find what province to use as its initializing source.
+		const auto& sourceProvince = determineProvinceSource(impProvinces, sourceWorld);
+		if (!sourceProvince)
+		{
+			continue; // MISMAP, or simply have mod provinces loaded we're not using.
+		}
+		else
+		{
+			province.second->initializeFromImperator(sourceProvince->second, cultureMapper, religionMapper);
+		}
+		// And finally, initialize it.
+		counter++;
+	}
+	LOG(LogLevel::Info) << ">> " << sourceWorld.getProvinces().size() << " Imperator provinces imported into " << counter << " CK3 provinces.";
+}
+
+std::optional<std::pair<int, std::shared_ptr<ImperatorWorld::Province>>> CK3::World::determineProvinceSource(const std::vector<int>& impProvinceNumbers,
+	const ImperatorWorld::World& sourceWorld) const
+{
+	// determine ownership by province development.
+	std::map<int, std::vector<std::shared_ptr<ImperatorWorld::Province>>> theClaims; // owner, offered province sources
+	std::map<int, int> theShares;														// owner, development
+	int winner = -1;
+	auto maxDev = -1;
+
+	for (auto ck2ProvinceID : impProvinceNumbers)
+	{
+		const auto& impProvince = sourceWorld.getProvinces().find(ck2ProvinceID);
+		if (impProvince == sourceWorld.getProvinces().end())
+		{
+			continue; // Broken mapping, or loaded a mod changing provinces without using it.
+		}
+		auto owner = impProvince->second->getOwner();
+		theClaims[owner].push_back(impProvince->second);
+		theShares[owner] = lround(impProvince->second->getBuildingsCount());
+	}
+	// Let's see who the lucky winner is.
+	for (const auto& share : theShares)
+	{
+		if (share.second > maxDev)
+		{
+			winner = share.first;
+			maxDev = share.second;
+		}
+	}
+	if (winner == -1)
+	{
+		return std::nullopt;
+	}
+
+	// Now that we have a winning owner, let's find its largest province to use as a source.
```

Comment:
Not sure if I like this. What if the largest differs from all other, and the others are collectively 'weightier' than the largest? Shouldn't they matter more in that case?

Reply:
This code is from ck2toeu4. It files all owners separately and gives them their "shares". Largest one is chosen and then we link to the largest of his provinces, due to 1-to-1 ck2-eu4 province mapping. There is no others, as all the people involved are independent from each other.